### PR TITLE
Phylo.Tree to igraph Graph

### PR DESCRIPTION
**Pull request by iosonofabio**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Hello there, I am a core developer of `igraph` and `python-igraph` and would like to add a function to convert a `Bio.Phylo.Tree` object into an `igraph.Graph` for downstream analyses and visualisation.

I have not tested the changes yet but wanted to hear your opinion on the PR first.

### CODE REVIEW

File: CONTRIB.rst [hunk 1]
```diff
@@ -116,6 +116,7 @@ please open an issue on GitHub or mention it on the mailing list.
 - Erik Weßels <https://github.com/BioNij>
 - Erik Whiting <https://github.com/erik-whiting>
 - Fabian Egli <https://github.com/fabianegli>
+- Fabio Zanini <https://github.com/iosonofabio>
```
Comment (peterjc) [inline CONTRIB.rst:119]: You can add yourself to the NEWS file too, perhaps also add a line about the new functionality too?

Comment (iosonofabio) [reply CONTRIB.rst:119]: yep done

File: Tests/test_Phylo_igraph.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Copyright (C) 2009 by Eric Talevich (eric.talevich@gmail.com)
```
[the remark this answers is not in the document]
Comment (iosonofabio) [reply Tests/test_Phylo_igraph.py]: yep fixed

**Comment by iosonofabio** [comment]
@ntamas FYI

**Comment by iosonofabio** [comment]
I now added the option to store arbitrary `Clade` properties into vertex/edge attributes of the `igraph.Graph` object.

**Review by etal — APPROVED** [review]

File: Bio/Phylo/_utils.py [hunk 3]
```diff
@@ -82,6 +82,107 @@ def build_subgraph(graph, top):
...
+        import igraph as ig
+    except ImportError:
+        raise MissingPythonDependencyError(
+            "Install igraph if you want to use to_igraph."
+        ) from None
+
+    if ig.__version__ < "0.10":
+        raise MissingPythonDependencyError(
+            "Update igraph to 0.10 or later if you want to use to_igraph."
+        ) from None
```
Comment (ntamas) [inline Bio/Phylo/_utils.py:113]: `from None` is not needed here; we are not in an exception block.

File: Bio/Phylo/_utils.py [hunk 4]
```diff
@@ -82,6 +82,107 @@ def build_subgraph(graph, top):
...
+        # NOTE: Bio.Phylo stores both vertex and edge attributes in the
+        # same place as "clade" attributes. However, the root node can have
+        # a vertex attribute but no edge attribute, hence the code blocks are
+        # in different places
+        for attrname in vertex_attrs:
+            if hasattr(node, attrname):
+                # Sometimes, only some tree nodes (e.g. leaves) have
+                # a certain attribute, e.g. a name
+                if len(vertex_attrs[attrname]) == 0:
+                    vertex_attrs[attrname] = [None for x in range(n_nodes)]
```
Comment (ntamas) [inline Bio/Phylo/_utils.py:137]: `[None] * n_nodes` avoids the creation of a range and is probably more readable.

**Review by ntamas — APPROVED** [review]

File: Bio/Phylo/_utils.py (hunk 4 above)
Comment (iosonofabio) [reply Bio/Phylo/_utils.py]: done

File: Bio/Phylo/_utils.py (hunk 3 above)
Comment (iosonofabio) [reply Bio/Phylo/_utils.py]: done

**Comment by iosonofabio** [comment]
Thanks @ntamas for the suggestions, which I implemented now. I think if this passes CI it is ready to merge.

**Comment by peterjc** [comment]
We should have merged this in March - I've dealt with the minor conflicts in the news and contributor listing, and intend we squash-and-merge this once the CI has finished.

**Comment by peterjc** [comment]
Thank you all - this will be in Biopython 1.83

**Pull request merged**